source/fuzz/transformation_add_read_modify_write_atomic_instruction.cpp

```diff
@@ -0,0 +1,487 @@
+  message_.set_fresh_id(fresh_id);
+  message_.set_pointer_id(pointer_id);
+  message_.set_opcode(opcode);
+  message_.set_memory_scope_id(memory_scope_id);
+  message_.set_memory_semantics_id_1(memory_semantics_id_1);
+  message_.set_memory_semantics_id_2(memory_semantics_id_2);
+  message_.set_value_id(value_id);
+  message_.set_comparator_id(comparator_id);
+
+  *message_.mutable_instruction_to_insert_before() =
+      instruction_to_insert_before;
+}
+
+bool TransformationAddReadModifyWriteAtomicInstruction::IsApplicable(
+    opt::IRContext* ir_context,
+    const TransformationContext& transformation_context) const {
+  // The result id must be fresh.
+  if (!fuzzerutil::IsFreshId(ir_context, message_.fresh_id())) {
+    return false;
+  }
+
+  // The pointer must exist and have a type.
+  auto pointer = ir_context->get_def_use_mgr()->GetDef(message_.pointer_id());
+  if (!pointer || !pointer->type_id()) {
+    return false;
+  }
+
+  // Read-Modify-Write atomic instructions are valid only.
+  if (pointer->opcode() == SpvOpAtomicLoad ||
+      pointer->opcode() == SpvOpAtomicStore) {
+    return false;
+  }
+
+  // The type must indeed be a pointer type.
+  auto pointer_type = ir_context->get_def_use_mgr()->GetDef(pointer->type_id());
+  assert(pointer_type && "Type id must be defined.");
+  if (pointer_type->opcode() != SpvOpTypePointer) {
+    return false;
+  }
+  // We do not want to allow loading from null or undefined pointers, as it is
+  // not clear how punishing the consequences of doing so are from a semantics
+  // point of view.
+  switch (pointer->opcode()) {
+    case SpvOpConstantNull:
+    case SpvOpUndef:
+      return false;
+    default:
+      break;
+  }
+
+  // Determine which instruction we should be inserting before.
+  auto insert_before =
+      FindInstruction(message_.instruction_to_insert_before(), ir_context);
+  // It must exist, ...
+  if (!insert_before) {
+    return false;
+  }
+
+  if (!fuzzerutil::CanInsertOpcodeBeforeInstruction(pointer->opcode(),
+                                                    insert_before)) {
+    return false;
+  }
+
+  // The block we are inserting into needs to be dead, or else the pointee type
+  // of the pointer needs to be irrelevant.
+  if (!transformation_context.GetFactManager()->BlockIsDead(
+          ir_context->get_instr_block(insert_before)->id()) &&
+      !transformation_context.GetFactManager()->PointeeValueIsIrrelevant(
+          message_.pointer_id())) {
+    return false;
+  }
+
+  // Instruction must RMW be atomic instruction, Id must be suitable for
+  // instruction, Check the validity of (Value, second memory semantics id,
+  // comparator id) for instruction if found.
+  if (!IsOperandIdsValidForInstruction(pointer->opcode(), ir_context,
+                                       insert_before, pointer_type)) {
+    return false;
+  }
+
+  // Check the validity of memory scope.
```

Comment:
I don't think I understand the code below this point either. Are we checking `value_id` and `comparator_id` somewhere? Why are we checking `insert_before`? `insert_before` could be any instruction; it is not necessarily an atomic instruction. It is not the instruction we are creating.

Reply:
Oh I see. Please ignore this.